crates/wasm-smith/src/component.rs

```diff
@@ -0,0 +1,802 @@
+        todo!()
+    }
+
+    fn arbitrary_alias_section(&mut self, u: &mut Unstructured<'a>) -> Result<()> {
+        todo!()
+    }
+}
+
+#[derive(Debug)]
+enum Section<'a> {
+    Custom(CustomSection<'a>),
+    Type(TypeSection),
+    Import(ImportSection),
+    Func(FuncSection),
+    Core(CoreSection),
+    Component(ComponentSection),
+    Instance(InstanceSection),
+    Export(ExportSection),
+    Start(StartSection),
+    Alias(AliasSection),
+}
+
+#[derive(Debug)]
+struct CustomSection<'a> {
+    name: &'a str,
+    data: &'a [u8],
+}
+
+impl<'a> Arbitrary<'a> for CustomSection<'a> {
+    fn arbitrary(u: &mut Unstructured<'a>) -> Result<Self> {
+        let name = limited_str(1_000, u)?;
+        let data = u.arbitrary()?;
+        Ok(CustomSection { name, data })
+    }
+}
+
+#[derive(Debug)]
+struct TypeSection {
+    types: Vec<Type>,
+}
+
+#[derive(Clone, Debug)]
+enum Type {
+    Module(ModuleType),
+    Component(ComponentType),
+    Instance(InstanceType),
+    Func(FuncType),
+    Value(ValueType),
+    Compound(CompoundType),
+}
+
+#[derive(Clone, Debug)]
+struct ModuleType {
+    defs: Vec<ModuleTypeDef>,
+}
+
+#[derive(Clone, Debug)]
+enum ModuleTypeDef {
+    TypeDef(crate::core::Type),
+    Import(crate::core::Import),
+    Export(String, crate::core::EntityType),
+}
+
+#[derive(Clone, Debug)]
+enum Alias {
+    InstanceExport {
+        instance: u32,
+        name: String,
+        kind: InstanceExportAliasKind,
+    },
+    Outer {
+        count: u32,
+        i: u32,
+        kind: OuterAliasKind,
+    },
+}
+
+#[derive(Clone, Debug)]
+enum InstanceExportAliasKind {
+    Module,
+    Component,
+    Instance,
+    Func,
+    Value,
+    Table,
+    Memory,
+    Global,
+}
+
+#[derive(Clone, Debug)]
+enum OuterAliasKind {
+    Module,
+    Component,
+    Type,
+}
+
+#[derive(Clone, Debug)]
+struct ComponentType {
+    defs: Vec<ComponentTypeDef>,
+}
+
+#[derive(Clone, Debug)]
+enum ComponentTypeDef {
+    Instance(InstanceTypeDef),
+    Import(Import),
+}
+
+#[derive(Clone, Debug)]
+struct InstanceType {
+    defs: Vec<InstanceTypeDef>,
+}
+
+#[derive(Clone, Debug)]
+enum InstanceTypeDef {
+    Type(Type),
+    Export { name: String, ty: u32 },
+    Alias(Alias),
+}
+
+#[derive(Clone, Debug)]
+struct FuncType {
+    params: Vec<NamedType>,
+    result: ValueType,
+}
+
+#[derive(Clone, Debug)]
+struct NamedType {
+    name: String,
+    ty: ValueType,
+}
+
+#[derive(Clone, Debug)]
+enum ValueType {
+    Indexed(u32),
+    Unit,
+    Bool,
+    S8,
+    U8,
+    S16,
+    U16,
+    S32,
+    U32,
+    S64,
+    U64,
+    F32,
+    F64,
+    Char,
+    String,
+}
+
+#[derive(Clone, Debug)]
+enum CompoundType {
+    Record(RecordType),
+    Variant(VariantType),
+    List(ListType),
+    Tuple(TupleType),
+    Flags(FlagsType),
+    Enum(EnumType),
+    Union(UnionType),
+    Optional(OptionalType),
+    Expected(ExpectedType),
+    Named(NamedType),
```

Comment:
This is probably more of a spec question, but what is a named type at the spec level? (or why have named types?)

Also, this is notably missing resources which is a major complexity point in my mind, so I was curious if that was missing from the spec or intentionally missing from here

Reply:
I'm not totally sure. Semantically they're identical to the underlying type.

Reply:
The spec was updated earlier this week to remove named compound types in favor of being able to export types as a way to "name" them.

I haven't yet updated `wasm-encoder`, but we can probably remove this variant for now.

Reply:
Oh and Luke is working on getting resources and handles into the spec.